### DISCUSSION TITLE
collection/generic

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#271] [Announcements]
# collection/generic

本文详细讲解了 Java 泛型的概念、原理及应用技巧，为您展示了如何通过泛型提高代码的可重用性、类型安全和可读性。学习本文将帮助您更好地掌握 Java 泛型编程，提高编程效率与质量。

https://javabetter.cn/collection/generic.html

**Comment by IMZHEYA** [comment]


**Reply by AKA-K** [reply]
是的，应该是“我们可以向该集合中添加 Dog 类型的元素，也可以添加它的父类。但是，不能向其中添加 Cat 类型的元素，因为 Cat 不是 Dog 的父类。”

**Reply by AKA-K** [reply]
理解了半天快晕了，原来是笔误 = =

**Reply by YoruWong** [reply]
没有笔误，确实应该就是子类，因为子类对象无论如何都可以向上转型。

**Reply by chulinbai** [reply]
同上，没有笔误。

**Reply by chulinbai** [reply]
List<? super Dog> animals = new ArrayList<>();   
这句代码补充省略部分后，实际上是
List<? super Dog> animals = new ArrayList<Dog>();
animals能存储的还是Dog以及它的子类对象。
super和extends我们说的上界和下界指泛型类型的上下界，也就是实例化时new List<T>()中的T。当super时，T只能是Dog以及父类；当extends时，T只能时Dog以及子类。
在使用animals的时候，能存储的类型看实例化时的具体类型，与super和extends无关。

**Reply by chulinbai** [reply]
第三行打漏了。补充下
“”“
List<? super Dog> animals = new ArrayList<>();
这句代码补充省略部分后，实际上是
List<? super Dog> animals = new ArrayList<Dog>();
”“”

**Reply by chulinbai** [reply]
第三行
“”“
List<? super Dog> animals = new ArrayList<>();
这句代码补充省略部分后，实际上是
List<? super Dog> animals = new ArrayList<  Dog  >();
”“”
第五行
“”“
super和extends我们说的上界和下界指泛型类型的上下界，也就是实例化时new List< T >()中的T。当super时，T只能是Dog以及父类；当extends时，T只能时Dog以及子类。
”“”

**Comment by NotQuiet123** [comment]
感觉你们的解释都有点别扭，让人看得还是摸不着头脑。实际上我觉得还是二哥写得太浅显或者说太简单了，因为这本身是一个极其容易混淆的点，若写成这样
现在我们有一个 List<? super Dog> 集合，它的类型参数必须是 Dog 或其父类类型，也就是说这个集合里的类型参数的类型会是Dog、Animal或Object的其中一个。而我们可以向该集合中添加 Dog 类型的元素，也可以添加它的子类（这里要注意是添加，添加和刚才说的集合里的类型是不一样的东西）。但是，不能向其中添加 Cat 类型的元素，因为 Cat 不是 Dog 的子类。
我觉得误解就会少很多，如若可以在代码层面上辅助解释所谓添加和集合里类型的不同点，即
//  声明：animals 是一个元素类型为 "Dog 或其父类" 的列表
List<? super Dog> animals = new ArrayList<>(); // 实际类型被推断为 Dog
//  添加：只能放 Dog 及其子类（因为它们都能向上转型为 Dog/Animal/Object）
animals.add(new Dog());      //  OK
animals.add(new Bulldog());  //  OK (Bulldog extends Dog)
//  这样的操作能够保证列表本身的类型范围会更大，而添加进去的元素由于子类可向上转型而保证添加的元素不会超出列表本身的类型范围
那么我认为，这样下来绝对是没有疑惑点的，表述也是清楚明细的
